fix goroutine build up from connected notifications

vyzo opened the pull request:
As observed in the gateway.

Aggregates notifications and batch processes, while being careful not to block the event loop.

vyzo committed:
don't accumulate pending goroutines from new connections

vyzo committed:
use a semaphore for new peer notifications so that we don't block the event loop

vyzo committed:
emit new peer notification without holding the semaphore

vyzo committed:
do the notification in a goroutine

don't block the swarm while waiting for the semaphore.

aschmahmann commented on pubsub.go:
```diff
@@ -621,6 +614,43 @@ func (p *PubSub) processLoop(ctx context.Context) {
+	newPeers := p.newPeersPend
+	p.newPeersPend = make(map[peer.ID]struct{})
+
+	for pid := range newPeers {
```
Are we concerned that if this map is large we could block the event loop?

vyzo replied:
well, if it gets large enough to take a non trivial amount of time to process, we've got bigger problems! We are getting flooded with connections.

aschmahmann commented on pubsub.go:
```diff
@@ -621,6 +614,43 @@ func (p *PubSub) processLoop(ctx context.Context) {
+
+	default:
+		// contention, return and wait for the next notification without blocking the event loop
+		return
```
This seems like it might be ok, although is it a problem if there's contention for a long time?

vyzo replied:
i don't think so, at worst some peers wont be processed for a while; we can live with that.

aschmahmann reviewed (commented):
Left a couple of comments/questions, generally I get the changes and trust your instincts here.

Stebalien commented on pubsub.go:
```diff
@@ -260,6 +264,8 @@ func NewPubSub(ctx context.Context, h host.Host, rt PubSubRouter, opts ...Option
 		counter:               uint64(time.Now().UnixNano()),
 	}
 
+	ps.newPeersSema <- struct{}{}
```
nit: I generally prefer the pattern where _writing_ takes the lock, and reading unlocks. Just means you can avoid this initialization part.

Stebalien commented on pubsub.go:
```diff
@@ -621,6 +614,43 @@ func (p *PubSub) processLoop(ctx context.Context) {
+
+	newPeers := p.newPeersPend
+	p.newPeersPend = make(map[peer.ID]struct{})
+
```
Can we release the "lock" right here? There's no reason to hold it.

vyzo replied:
sure

vyzo replied:
done

Stebalien commented on pubsub.go:
```diff
@@ -231,7 +233,9 @@ func NewPubSub(ctx context.Context, h host.Host, rt PubSubRouter, opts ...Option
 		incoming:              make(chan *RPC, 32),
-		newPeers:              make(chan peer.ID),
+		newPeers:              make(chan struct{}, 1),
+		newPeersSema:          make(chan struct{}, 1),
```
I'd really just use a lock here. We're just protecting reads/writes to a single map.

vyzo replied:
Not really -- we can block the event loop indefinitely if there is a lot of contention in the lock; better to use a semaphore that allows us to try lock.

vyzo committed:
use a priority lock instead of a semaphore

so that there is no case of infinite accumulation of pending peers in the queue.
also adds a connectedness check before adding the peer.

vyzo committed:
batch process dead peer notifications

vyzo committed:
don't respawn writer if we fail to open a stream; declare it a peer error

vyzo commented:
rebased on master and also:
- added batching for dead peer notifications 
- fixed the handling of errors from opening new streams so that we can no longer go into loops of respawning writers because we failed to open a stream.

vyzo committed:
release priority locks early when handling batches

Stebalien commented on comm.go:
```diff
@@ -88,22 +87,29 @@ func (p *PubSub) handleNewStream(s network.Stream) {
-		if err == ms.ErrNotSupported {
-			ch = p.newPeerError
-		} else {
-			ch = p.peerDead
```
Why are we getting rid of this? Can't we fail to create a new stream if, e.g., the connection is dead? This will leave us without a peer if we try to use the wrong connection.

vyzo replied:
it can lead to some very nasty "peer declared dead; respawning writer" loops, which have been reported by lotus users in the wild.

Not sure how to properly handle this situation, let me give it a bit more thought -- it's subtle and what we have right now is very problematic.

vyzo replied:
and that's why I removed it.

Stebalien replied:
Ah, I see. Yeah. This is a crap situation. I'm just concerned that this will lead to the _opposite_ problem, but I think this is fine for now. But we should put in a better fix later.

vyzo replied:
Yeah, I am concerned about it too. I'll open a follow up issue so that we don't forget.

vyzo replied:
follow up issue in https://github.com/libp2p/go-libp2p-pubsub/issues/433

Stebalien commented on pubsub.go:
```diff
@@ -621,6 +590,83 @@ func (p *PubSub) processLoop(ctx context.Context) {
+			log.Debugf("peer declared dead but still connected; respawning writer: %s", pid)
+			messages := make(chan *RPC, p.peerOutboundQueueSize)
+			messages <- p.getHelloPacket()
+			go p.handleNewPeer(p.ctx, pid, messages)
```
You know, _this_ might have been the source of backing up handleNewPeer goroutines... It's possible that we were seeing "unresponsive" peers that still appeared to be connected?

Maybe we need some form of connection blacklist? Or maybe a bit of backoff?

vyzo replied:
yeah, same problem as above; we need a more principled approach.

Stebalien replied:
:sob:
